internal/charmstore/store.go

```diff
@@ -526,9 +526,16 @@ func (s *Store) AddCharm(c charm.Charm, p AddParams) (err error) {
 	if p.URL.URL.Series == "bundle" || p.URL.URL.User == "" || p.URL.URL.Revision == -1 {
 		return errgo.Newf("charm added with invalid id %v", &p.URL.URL)
 	}
-	if p.URL.URL.Series == "" && len(c.Meta().Series) == 0 {
-		return errgo.Newf("charm added without series %v", &p.URL.URL)
+	if len(c.Meta().Series) == 0 {
```

Comment:
I think my initial confusion arose from the fact we're switching on the meta series, not the uploaded series.

An alternative (hopefully equivalent) phrasing, using the uploaded series to drive the logic, which fits better with the way my brain sees it:

    if id.Series == "" {
         if len(c.Meta().Series) == 0 {
             return errgo.WithCausef(nil, params.ErrBadRequest, "charm %v added without any supported series", &p.URL.URL)
         }
     } else if len(c.Meta.Series) != 0 && !contains(id.Series, c.Meta().Series) {
         return errgo.WithCausef(nil, params.ErrForbidden, "%s not listed in charm metadata", p.URL.URL.Series)
     }

Take it or leave it.